### PR TITLE
RavenDB-15020 : Additional time series functions

### PR DESCRIPTION
**Pull request by aviv86**
https://issues.hibernatingrhinos.com/issue/RavenDB-15020

### CODE REVIEW

File: src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs [hunk 1]
```diff
@@ -145,7 +145,7 @@ public class TimeSeriesAggregationResult<T> : TimeSeriesQueryStreamResultBase<Ti
     public class TimeSeriesRangeAggregation : IPostJsonDeserialization, ITimeSeriesQueryStreamEntry
     {
         public long[] Count;
-        public double[] Max, Min, Last, First, Average, Sum;
+        public double[] Max, Min, Last, First, Average, Sum, Percentile, Slope;
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs]: Wonder if we should call it `Derivative` (or maybe `Change`?) instead of `Slope`.

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+                        break;
+                    case AggregationType.Last:
+                        _values[i] = val;
+                        break;
+                    case AggregationType.Count:
+                        break;
+                    case AggregationType.Percentile:
+                        Debug.Assert(_rankedValues != null);
+                        var currentRankedList = _rankedValues[i];
+                        var loc = currentRankedList.BinarySearch(val);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: I would prefer to use `SortedDictionary` with a comparer to handle duplicates.

File: src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs]: Isn't slope the right math term?

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: Agreed, also, and `insert` here will cause us to move a lot of values, potentially.

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+            if (scale.HasValue == false) 
+                return _finalValues = _values;
+
+            return _finalValues = _values.Select(x => x * scale.Value);
+        }
+
+        private IEnumerable<double> GetPercentile(double scale)
+        {
+            foreach (var list in _rankedValues)
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: This rounds down, right? Do we need some manner to adjust the value based on the distance?
So if we have [1,2,4] and we ask for 75% percentile, we'll get [2] here. But that actually represent the 50%.
Should we adjust that so it would be 3 as the result?

Comment (aviv86) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: No actually we round up : `index = (int)Math.Ceiling(_percentileFactor * list.Count)`
so in your example we'll get `4` as the result. (`2` is wrong by any definition)

There are different definitions for percentile and methods for calculating percentile.
By the definition that I chose to implement the result is always a part of the data set.
So if we want to adjust the result to be `3` we need to use a different method..

I wrote more about it in the issue page : https://issues.hibernatingrhinos.com/issue/RavenDB-15020#focus=Comments-67-51024.0-0

File: src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs]: I think `Slope` is the right math term. 
`Change` is inaccurate - we are measuring **the rate** of change
and `Derivative` is the rate of change when the distance (delta X) tends towards 0, no?

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: Okay, that make sense.
If it isn't hard, let's go with extrapolation. Note that we need to document this.

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs (hunk 2 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: done

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs (hunk 3 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: done

**Comment by aviv86** [comment]
- Added `StandardDeviation` support
- Made some refactoring in `TimeSeriesAggregation`  (introduced `ITimeSeriesAggregation`)
- Addressed PR comments

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            for (int i = 0; i < values.Length; i++)
+            {
+                _values[i] += values[i];
+                _count[i]++;
+            }
+        }
+
+        IEnumerable<double> ITimeSeriesAggregation.GetFinalValues(DateTime? @from, DateTime? to, double? scale = null)
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs:84]: Can you check the warning?
```
src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs#L83
The default value specified for parameter 'scale' will have no effect because it applies to a member that is used in contexts that do not allow optional arguments
```

Probably just remove `= null` from `scale` here. It is an explicit interface impl.

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                if (i < oldCount)
+                {
+                    _first[i] = val.First;
+                }
+
+                _values[i] = val.Last;
+            }
+        }
+
+        IEnumerable<double> ITimeSeriesAggregation.GetFinalValues(DateTime? @from, DateTime? to, double? scale = null)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs]: Warning here. Explicit interface impl.

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+        public string Name { get; }
+
+        public int NumberOfValues { get; }
+
+        public AggregationType Aggregation { get; }
+
+    }
+
+    public class TimeSeriesAggregationBase
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: abstract?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs (hunk 4 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs:84]: done

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs (hunk 5 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs]: done

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs (hunk 6 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs]: done

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs [hunk 7]
```diff
@@ -145,7 +145,7 @@ public class TimeSeriesAggregationResult<T> : TimeSeriesQueryStreamResultBase<Ti
     public class TimeSeriesRangeAggregation : IPostJsonDeserialization, ITimeSeriesQueryStreamEntry
     {
         public long[] Count;
-        public double[] Max, Min, Last, First, Average, Sum;
+        public double[] Max, Min, Last, First, Average, Sum, Percentile, Slope, StandardDeviation;
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: Should we maybe shorten this to StdDev?

File: src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            }
+
+            var oldCount = _values.Count;
+            InitValuesIfNeeded(values.Length);
+
+            for (int i = 0; i < values.Length; i++)
+            {
+                var val = values[i];
+
+                if (i < oldCount)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs]: for consistency with above `oldCount <= i`

**Review by karmeli87 — APPROVED** [review]
Very Nice! I like the `TimeSeriesAggregation` abstraction :-)

File: src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs (hunk 7 above)
Comment (ayende) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: yes

Comment (ppekrol) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: No, then Average should be Avg, etc. Either use full names of abbreviations. Not both.

Comment (ppekrol) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: Actually it is already mixed, nvm.

Comment (ml054) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: May I ask what's definition used for slope?

Comment (aviv86) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: The rate of change within a range (only works with group by)
i.e. `slope = dY/dX`, dY = `range.LastVal - range.FirstVal`, dX = `(range.End - range.Start).TotalMilliseconds`

Comment (ml054) [reply src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:148]: ok I thought  it uses linear regression
